File: algorithms/1_easy/P/02848_points_that_intersect_with_cars/solution.cpp
class Solution {
public:
    int numberOfPoints(vector<vector<int>>& nums) {
        sort(nums.begin(), nums.end(), [](const auto& a, const auto& b) {
            return a[0] < b[0];
        });

        int count{0};
        int prev_right_point{numeric_limits<int>::min()};

        for (const auto& point : nums) {
            if (prev_right_point >=  point[1]) {
                continue;
            }

            if (prev_right_point >=  point[0]) {
                count += point[1] - prev_right_point;
                prev_right_point = point[1];
            } else {
                count += point[1] - point[0] + 1;
                prev_right_point = point[1];
            }
        }

        return count;
    }
};

File: algorithms/1_easy/P/02848_points_that_intersect_with_cars/solution.rs
impl Solution {
    pub fn number_of_points(mut nums: Vec<Vec<i32>>) -> i32 {
        nums.sort_unstable_by(|a, b| a[0].cmp(&b[0]));
        let mut count = 0_i32;
        let mut prev_right_point = i32::MIN;

        for point in nums.iter() {
            if prev_right_point >=  point[1] {
                continue;
            }

            if prev_right_point >=  point[0] {
                count += point[1] - prev_right_point;
                prev_right_point = point[1];
            } else {
                count += point[1] - point[0] + 1;
                prev_right_point = point[1];
            }
        }

        count
    }
}
